### PR TITLE
Identify refactor

### PR DESCRIPTION
**Pull request by profxj**
Refactors the `identify` GUI for the fitting and wavecalib updates.

Adds more docs and unit tests.

I don't intend to run the Dev Suite as none of the internal files were touched.
But I will run unit tests.

### CODE REVIEW

**Comment by profxj** [comment]
================ 251 passed, 816 warnings in 634.52s (0:10:34) =================

**Review by rcooke-ast — APPROVED** [review]
This looks good to me - Most of my suggestions are a matter of preference, so I'm approving... Did you play around with the "ghost spectrum" at all (i.e. was it useful/helpful)?  If so, I wonder if this should be added to the docs...

File: pypeit/core/gui/identify.py [hunk 1]
```diff
@@ -245,7 +257,7 @@ def initialise(cls, arccen, slits, slit=0, par=None, wv_calib_all=None, wavelim=
         residcmap = LinearSegmentedColormap.from_list("my_list", lflag_color, N=len(lflag_color))
         resres = axres.scatter(detns, np.zeros(detns.size), marker='x',
                                c=np.zeros(detns.size), cmap=residcmap, norm=Normalize(vmin=0.0, vmax=3.0))
-        axres.axhspan(-0.1, 0.1, alpha=0.5, color='grey')  # Residuals of 0.1 pixels
+        axres.axhspan(-1*pxtoler, pxtoler, alpha=0.5, color='grey')  # Residuals of 0.1 pixels
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/gui/identify.py:260]: removed

File: pypeit/core/gui/identify.py [hunk 2]
```diff
@@ -422,11 +436,12 @@ def draw_lines(self):
             xnorm = self._fitdict['xnorm']
 
             # Calculate the estimated wavelength of the detections
-            plotx = utils.func_val(self._fitdict['fitc'],
-                                   pixel_fit / xnorm,
-                                   self._fitdict["function"],
-                                   minx=self._fitdict['fmin'],
-                                   maxx=self._fitdict['fmax'])
+            plotx = self._fitdict['full_fit'].eval(pixel_fit / xnorm)
+            #plotx = utils.func_val(self._fitdict['fitc'],
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/gui/identify.py:440]: gone

File: pypeit/core/gui/identify.py [hunk 3]
```diff
@@ -643,40 +660,52 @@ def store_solution(self, final_fit, master_dir, binspec, rmstol=0.15,
...
-                          msgs.newline() + templates.outpath + msgs.newline() + msgs.newline() +
-                          "Please consider sending your solution to the PypeIt team!" + msgs.newline())
+                # Arxiv solution
+                #outroot = templates.pypeit_identify_record(final_fit, binspec, specname, gratname, dispangl, outdir=master_dir)
+                wavelengths = self._fitdict['full_fit'].eval(np.arange(self.specdata.size) /
+                                                             (self.specdata.size - 1))
+                templates.write_template(wavelengths, self.specdata, binspec,
+                                         './', 'wvarxiv.fits')
+                msgs.info("\nYour arxiv solution has been written to wvarxiv.fits")
+                #msgs.info("\nYour wavelength solution has been stored here:" + msgs.newline() +
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/gui/identify.py:695]: yes, we should.  In the docs.  
But, to be honest, I don't know what to instruct them to do. ;)

Can you remind me?

File: pypeit/scripts/identify.py [hunk 4]
```diff
@@ -29,6 +29,10 @@ def parse_args(options=None, return_parser=False):
     parser.add_argument("--slit", type=int, default=0, help="Which slit to load for wavelength calibration")
     parser.add_argument("--det", type=int, default=1, help="Detector index")
     parser.add_argument("--rmstol", type=float, default=0.1, help="RMS tolerance")
+    parser.add_argument("--fwhm", type=float, default=4., help="FWHM for line finding")
+    parser.add_argument("--pixtol", type=float, default=0.1, help="Pixel tolerance for Auto IDs")
+    parser.add_argument('--test', default=False, action='store_true',
+                        help="Unit tests?")
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/scripts/identify.py:35]: good suggestion.  Done.

File: pypeit/core/gui/identify.py [hunk 5]
```diff
@@ -643,40 +660,52 @@ def store_solution(self, final_fit, master_dir, binspec, rmstol=0.15,
             Grating name
         dispangl : str
             Dispersor Angle (expressed as a name/string)
+        force_yes : bool
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/gui/identify.py:663]: done

File: pypeit/core/gui/identify.py [hunk 6]
```diff
@@ -80,6 +79,7 @@ def __init__(self, canvas, axes, spec, specres, detns, line_lists, par, lflag_co
             Spatial ID corresponding to slit
         wv_calib : :obj:`dict`, None, optional
             If a best-fitting solution exists, and you wish to load it, provide the wv_calib dictionary.
+        pxtoler : float, optional
```
Comment (deborape) [inline pypeit/core/gui/identify.py:82]: Please, add the description for `pxtoler`

File: pypeit/core/gui/identify.py [hunk 7]
```diff
@@ -190,6 +196,10 @@ def initialise(cls, arccen, slits, slit=0, par=None, wv_calib_all=None, wavelim=
             If a best-fitting solution exists, and you wish to load it, provide the wv_calib dictionary.
         wavelim : :obj:`list`, None, optional
             A two element list containing the desired minimum and maximum wavelength of the linelist
+        test : bool, optional
+            If True, this is a unit test
+        fwhm : float, optional
+            FWHM of arc lines in pixels
```
Comment (deborape) [inline pypeit/core/gui/identify.py:202]: `nonlinear_counts` and `pxtoler` seem missing in this doc

File: pypeit/core/gui/identify.py [hunk 8]
```diff
@@ -643,40 +655,52 @@ def store_solution(self, final_fit, master_dir, binspec, rmstol=0.15,
             Grating name
         dispangl : str
             Dispersor Angle (expressed as a name/string)
+        force_save : bool
+            Force save
```
Comment (deborape) [inline pypeit/core/gui/identify.py:659]: `wvcalib` is missing in this doc

**Review by deborape — APPROVED** [review]
Looks good... and cool! Approved!

File: pypeit/core/gui/identify.py (hunk 6 above)
Comment (profxj) [reply pypeit/core/gui/identify.py:82]: done

File: pypeit/core/gui/identify.py (hunk 7 above)
Comment (profxj) [reply pypeit/core/gui/identify.py:202]: added

File: pypeit/core/gui/identify.py (hunk 8 above)
Comment (profxj) [reply pypeit/core/gui/identify.py:659]: added

**Comment by profxj** [comment]
merging

**Pull request merged**